### PR TITLE
Added support for regref transforms

### PR DESCRIPTION
**Pull request by josh146**
At the request of @smite, both the Blackbird spec and the Blackbird Python parser now have support for classical processing of measured modes/RegRefTransforms.

This is specified using the following syntax:
```python
with device(shots=100):
    S2gate(0.543, 0.123) | [0, 1]
    MeasureX | 0
    MeasureP | 1
    Xgate((q0+q1)/sqrt(2)) | 2
```

Any arithmetic operation supported by Blackbird can be used as a register transform.

Variables of the form `qX`, where `X` is an integer, are now **reserved**, and cannot be used in any variable declarations. Attempting to do the following:

```python
int q3 = 5
```

will result in a `NameError: Variable name q3 is reserved for register references`.

To enable symbolic functions to be represented and returned by the parser, Blackbird now depends on SymPy.

**Note:** As the grammar was modified and recompiled to allow this new behaviour, this makes this PR look significantly bigger than it is (since a lot of autogenerated files were re-generated).

### CODE REVIEW

File: .coveragerc [hunk 1]
```diff
@@ -1,6 +1,8 @@
 # .coveragerc to control coverage.py
 [run]
 source = blackbird_python/blackbird
+omit =
+    blackbird_python/blackbird/blackbird*.py
```
Comment (josh146) [inline .coveragerc]: These are the autogenerated parser files, which are very hard to test directly so are omitted from coverage.

Anything with filename matching `blackbird*.py` can also be ignored from this PR.

File: blackbird_python/blackbird/auxiliary.py [hunk 2]
```diff
@@ -169,7 +174,13 @@ def _expression(expr):
         if expr.TIMES():
             return np.prod([_expression(a), _expression(b)], axis=0)
         if expr.DIVIDE():
-            return np.prod([_expression(a), np.power(_expression(b), -1)], axis=0)
+            a = _expression(a)
+            b = _expression(b)
+
+            if isinstance(b, int):
+                b = float(b)
```
Comment (josh146) [inline blackbird_python/blackbird/auxiliary.py]: This was out of scope of the PR, oops. Noticed that numpy would complain if `b` was an integer, and the power was -1.

File: blackbird_python/blackbird/tests/test_listener.py [hunk 3]
```diff
@@ -249,6 +258,37 @@ def test_device_kwarg(self, parse_input):
...
+
+        rrt = RegRefTransform(f)
+
+        assert rrt.func_str == "0.707106781186547*q0 - 0.707106781186547*q2"
+        assert set(rrt.regrefs) == {0, 2}
+
+        if rrt.regrefs == [0, 2]:
+            assert np.allclose(rrt.func(0.543, -0.432), (0.543+0.432)/np.sqrt(2))
+
+        if rrt.regrefs == [2, 0]:
```
Comment (co9olguy) [inline blackbird_python/blackbird/tests/test_listener.py]: Are we expecting stochastic behaviour here?

File: blackbird_python/blackbird/__init__.py [hunk 4]
```diff
@@ -63,6 +68,12 @@
   This class can be sub-classed, to create more advanced Blackbird listeners
   that perform actions (i.e. such as simulations) upon parsing the tree.
```
Comment (co9olguy) [inline blackbird_python/blackbird/__init__.py]:
```suggestion
  that perform actions (e.g., simulations) upon parsing the tree.
```

File: blackbird_python/blackbird/auxiliary.py (hunk 2 above)
Comment (co9olguy) [reply blackbird_python/blackbird/auxiliary.py]: Ok, can you make issue so we remember to fix later?

File: blackbird_python/blackbird/tests/test_listener.py (hunk 3 above)
Comment (josh146) [reply blackbird_python/blackbird/tests/test_listener.py]: Yep, this seems to be an issue with SymPy. The order in which it returns the variables of a symbolic function are stochastic (and this is discussed on Stack Exchange).

In this particular case, sometimes it returns `(q[0], q[2])` as the free variables, other times it returns `(q[2], q[0])`.

File: blackbird_python/blackbird/auxiliary.py (hunk 2 above)
Comment (josh146) [reply blackbird_python/blackbird/auxiliary.py]: :+1:

**Comment by co9olguy** [comment]
Let's leave a comment about this in the code then

File: blackbird_python/blackbird/tests/test_listener.py (hunk 3 above)
Comment (josh146) [reply blackbird_python/blackbird/tests/test_listener.py]:
```suggestion
        # Note: the order in SymPy returns the variables of a symbolic function are
        # stochastic and may change at runtime. In this particular case, sometimes it returns
        # (q[0], q[2]) as the free variables, other times it returns (q[2], q[0]).
        #
        # This is not an issue during usage, as the Blackbird parser will always know the
        # correct order to apply the register arguments, but must be taken into account
        # during test assertions.
        if rrt.regrefs == [2, 0]:
```

**Pull request merged**